File: webapp/js/dockerrun.js
if(ws==undefined)var ws;
if(ws)ws.close();
ws = new WebSocket("ws://localhost:" + global.backendPort + "/web/app/events");
//通信が接続された場合
ws.onopen = function(e) { 
    console.log("Setup web socket:", ws);
    ws.send(JSON.stringify({
        "event": "dockerrun"
    }))  
    ws.onmessage = function (event) {
        console.log(event.data);
        const eObj = JSON.parse(event.data);
        console.log(eObj);
        const data = JSON.parse(eObj.event);
        console.log(data)
        if(data.eventName == "index-view") {
            $(".content-wrapper").html(NNSH_decodeHTML(data.view.view_data));
            return
        }
        else if(data.eventName == "log") {
            document.getElementById("console-output").innerHTML += "\n" + data.log;
            $('#console-output').animate({scrollTop: $('#console-output')[0].scrollHeight}, 'fast');
            return
        }
        else if(data.eventName == "dockerrun-load") {
            
            $("#name").val(data.runInputData.name);
            $("#dit").val(data.runInputData.dit);
            $("#port").val(data.runInputData.port);
            $("#option").val(data.runInputData.option);
            $("#dirname").val(data.runInputData.dirname);
            $("#dirname2").val(data.runInputData.dirname2);
            $("#dirname_a").val(data.runInputData.dirname_a);
            $("#dirname_a2").val(data.runInputData.dirname_a2);
            $("#dirname_b").val(data.runInputData.dirname_b);
            $("#dirname_b2").val(data.runInputData.dirname_b2);
            $("#dirname_c").val(data.runInputData.dirname_c);
            $("#dirname_c2").val(data.runInputData.dirname_c2);
            $("#dirname_d").val(data.runInputData.dirname_d);
            $("#dirname_d2").val(data.runInputData.dirname_d2);
            $("#option2").val(data.runInputData.option2);
            //ないかもしれないのでラスト
            $("#docker").val(data.runInputData.docker);
            return
        }
        if(data.eventName == "dockerrun-saveinput") {

            try {$('#inputlist').columns('destroy');} catch(e) {}

            if(data.runInputlist) {
                $('#inputlist').columns({data:data.runInputlist,
                    schema: [
                        {"header":"NAME", "key":"name", "template":'<a href="#" onclick="loadRunInput(\'{{name}}\')">{{name}}</a>'},
                        {"header":"", "key":"cmd"},
                        {"header":" ", "key":"", "template":'<input type="button" onclick="deleteRunInput(\'{{name}}\')" value="DELETE" />'},
                ]
                });
            }
            return;
        }
        if(data.eventName == "dockerrun-delete") {

            try {$('#inputlist').columns('destroy');} catch(e) {}

            if(data.runInputlist) {
                $('#inputlist').columns({data:data.runInputlist,
                    schema: [
                        {"header":"NAME", "key":"name", "template":'<a href="#" onclick="loadRunInput(\'{{name}}\')">{{name}}</a>'},
                        {"header":"", "key":"cmd"},
                        {"header":" ", "key":"", "template":'<input type="button" onclick="deleteRunInput(\'{{name}}\')" value="DELETE" />'},
                ]
                });
            }
            return;
        }
        $.each(data.image, function (key, val) {
            $('#docker').append($('<option>').val(val.ID).text(val.Repository+"("+val.ID+")   "+ val.CreatedAt));
        });
        console.log(data.runInputlist);
        if(data.runInputlist) {
            $('#inputlist').columns({data:data.runInputlist,
                schema: [
                    {"header":"NAME", "key":"name", "template":'<a href="#" onclick="loadRunInput(\'{{name}}\')">{{name}}</a>'},
                    {"header":"", "key":"cmd"},
                    {"header":" ", "key":"", "template":'<input type="button" onclick="deleteRunInput(\'{{name}}\')" value="DELETE" />'},
            ]
            });
        }
    };
};
 
//エラーが発生した場合
ws.onerror = function(error) { };
 
//メッセージを受け取った場合
ws.onmessage = function(e) { };
  
//通信が切断された場合
ws.onclose = function() { };

//ページが読み込み完了した場合
$(document).ready(function () {

    $("#submit_button").click(function() {
        var fromData = formToData($("#frm"));
        fromData['event'] = "dockerrun-excute"

        ws.send(JSON.stringify(fromData))
        console.log("Send success!!");
    });
    $("#submit_button2").click(function() {
        var fromData = formToData($("#frm"));
        if($("#name").val() == "") {
            alert("require name")
            return;
        }


        fromData['event'] = "dockerrun-saveinput"

        ws.send(JSON.stringify(fromData))
        console.log("Send success!!");
    });

    $("#select_dir").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname").val(this.files[i].path)
          }
    });
    $("#select_dir2").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_a").val(this.files[i].path)
          }
    });
    $("#select_dir3").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_b").val(this.files[i].path)
          }
    });
    $("#select_dir4").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_c").val(this.files[i].path)
          }
    });
    $("#select_dir5").on('change', function() {
        for (i = 0; i < this.files.length; i++) {
            $("#dirname_d").val(this.files[i].path)
          }
    });
});
function loadRunInput(name) {

    ws.send(JSON.stringify({
        "event": "dockerrun-load",
        "name":name
    }));
}
function deleteRunInput(name) {

    ws.send(JSON.stringify({
        "event": "dockerrun-delete",
        "name":name
    }));
}

File: mylib/socket/dockerrun/dockerrun.go
package dockerrun

import (
	"bytes"
	"encoding/json"
	"fmt"

	"github.com/Equanox/gotron"
	"github.com/uieviiru2/mylib/docker"
	"github.com/uieviiru2/mylib/jsonedit"
)

type dockerRunStruct struct {
	Docker    string `json:"docker"`
	Name      string `json:"name"`
	Dit       string `json:"dit"`
	Port      string `json:"port"`
	DirName   string `json:"dirname"`
	DirName2  string `json:"dirname2"`
	DirNameA  string `json:"dirname_a"`
	DirNameA2 string `json:"dirname_a2"`
	DirNameB  string `json:"dirname_b"`
	DirNameB2 string `json:"dirname_b2"`
	DirNameC  string `json:"dirname_c"`
	DirNameC2 string `json:"dirname_c2"`
	DirNameD  string `json:"dirname_d"`
	DirNameD2 string `json:"dirname_d2"`
	Option    string `json:"option"`
	Option2   string `json:"option2"`
	Cmd       string `json:"cmd"`
}

func Use(window *gotron.BrowserWindow) {

	window.On(&gotron.Event{Event: "dockerrun"}, func(bin []byte) {

		output := jsonedit.Val("eventName", "dockerrun")
		jDockerImage := docker.Image(window)
		output = jsonedit.Con(output, jDockerImage)
		jServerfiles := docker.RunInputfiles()
		output = jsonedit.Con(output, jServerfiles)

		window.Send(&gotron.Event{Event: jsonedit.End(output)})

	})
	window.On(&gotron.Event{Event: "dockerrun-excute"}, func(bin []byte) {

		var d dockerRunStruct

		b := []byte(bin)
		//buf := bytes.NewBuffer(b)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(string(b))
		fmt.Println(d.Docker)
		fmt.Println(d.Name)
		fmt.Println(d.Port)
		fmt.Println(d.DirName)
		fmt.Println(d.DirName2)
		fmt.Println(d.DirNameA)
		fmt.Println(d.DirNameA2)
		fmt.Println(d.DirNameB)
		fmt.Println(d.DirNameB2)
		fmt.Println(d.DirNameC)
		fmt.Println(d.DirNameC2)
		fmt.Println(d.DirNameD)
		fmt.Println(d.DirNameD2)
		fmt.Println(d.Option)
		if docker.IsWindows() {
			docker.Run(d.Docker, d.Name, d.Dit, d.Port, d.DirName, d.DirName2, d.DirNameA, d.DirNameA2, d.DirNameB, d.DirNameB2, d.DirNameC, d.DirNameC2, d.DirNameD, d.DirNameD2, d.Option, d.Option2, window)
		} else {
			docker.RunMac(d.Docker, d.Name, d.Dit, d.Port, d.DirName, d.DirName2, d.DirNameA, d.DirNameA2, d.DirNameB, d.DirNameB2, d.DirNameC, d.DirNameC2, d.DirNameD, d.DirNameD2, d.Option, d.Option2, window)
		}
	})
	window.On(&gotron.Event{Event: "dockerrun-saveinput"}, func(bin []byte) {

		var d dockerRunStruct

		b := []byte(bin)
		//buf := bytes.NewBuffer(b)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(string(b))
		fmt.Println(d.Docker)
		fmt.Println(d.Name)
		fmt.Println(d.Port)
		fmt.Println(d.DirName)
		fmt.Println(d.DirName2)
		fmt.Println(d.DirNameA)
		fmt.Println(d.DirNameA2)
		fmt.Println(d.DirNameB)
		fmt.Println(d.DirNameB2)
		fmt.Println(d.DirNameC)
		fmt.Println(d.DirNameC2)
		fmt.Println(d.DirNameD)
		fmt.Println(d.DirNameD2)
		fmt.Println(d.Option)
		d.Cmd = makeRunStr(d)

		bytes, _ := json.Marshal(d)
		docker.SaveRunInputfile(d.Name, string(bytes), window)

		output := jsonedit.Val("eventName", "dockerrun-saveinput")
		jServerfiles := docker.RunInputfiles()
		output = jsonedit.Con(output, jServerfiles)
		window.Send(&gotron.Event{Event: jsonedit.End(output)})

	})
	window.On(&gotron.Event{Event: "dockerrun-load"}, func(bin []byte) {
		var d dockerRunStruct
		b := []byte(bin)
		buf := bytes.NewBuffer(b)
		fmt.Println(buf)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(d.Name)
		jServer := docker.LoadRunInputfile(d.Name)

		output := jsonedit.Val("eventName", "dockerrun-load")
		output = jsonedit.Con(output, jsonedit.On("runInputData", jServer))
		fmt.Println("output")
		fmt.Println(output)
		window.Send(&gotron.Event{Event: jsonedit.End(output)})
	})
	window.On(&gotron.Event{Event: "dockerrun-delete"}, func(bin []byte) {
		var d dockerRunStruct
		b := []byte(bin)
		buf := bytes.NewBuffer(b)
		fmt.Println(buf)
		if err := json.Unmarshal(b, &d); err != nil {
			// ...
		}
		fmt.Println(d.Name)
		docker.DeleteRunInputfile(d.Name, window)

		output := jsonedit.Val("eventName", "dockerrun-delete")
		jServerfiles := docker.RunInputfiles()
		output = jsonedit.Con(output, jServerfiles)
		window.Send(&gotron.Event{Event: jsonedit.End(output)})

	})
}
func makeRunStr(d dockerRunStruct) string {
	cmd := "docker run " + d.Dit + " --name " + d.Name
	if d.DirName != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirName) + ":/" + d.DirName2
	}
	if d.DirNameA != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameA) + ":/" + d.DirNameA2
	}
	if d.DirNameB != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameB) + ":/" + d.DirNameB2
	}
	if d.DirNameC != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameC) + ":/" + d.DirNameC2
	}
	if d.DirNameD != "" {
		cmd += " -v " + docker.ChangeDockerPath(d.DirNameD) + ":/" + d.DirNameD2
	}
	if d.Port != "" {
		cmd += " -p " + d.Port
	}
	if d.Option != "" {
		cmd += " " + d.Option
	}
	cmd += " " + d.Docker

	if d.Option2 != "" {
		cmd += " " + d.Option2
	}
	return cmd
}
